File: core/src/com/jediterm/terminal/Terminal.java
package com.jediterm.terminal;

import com.jediterm.core.Color;
import com.jediterm.core.util.CellPosition;
import com.jediterm.core.util.TermSize;
import com.jediterm.terminal.emulator.mouse.MouseFormat;
import com.jediterm.terminal.emulator.mouse.MouseMode;
import com.jediterm.terminal.model.StyleState;
import com.jediterm.terminal.model.TerminalApplicationTitleListener;
import com.jediterm.terminal.model.TerminalResizeListener;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.UnsupportedEncodingException;
import java.util.List;

/**
 * Executes terminal commands interpreted by {@link com.jediterm.terminal.emulator.Emulator}, receives text
 *
 * @author traff
 */
public interface Terminal {
  void resize(@NotNull TermSize newTermSize, @NotNull RequestOrigin origin);

  void beep();

  void backspace();

  void horizontalTab();

  void carriageReturn();

  void newLine();

  void mapCharsetToGL(int num);

  void mapCharsetToGR(int num);

  void designateCharacterSet(int tableNumber, char ch);

  void setAnsiConformanceLevel(int level);

  void writeDoubleByte(char[] bytes) throws UnsupportedEncodingException;

  void writeCharacters(String string);

  int distanceToLineEnd();

  void reverseIndex();

  void index();

  void nextLine();

  void fillScreen(char c);

  void saveCursor();

  void restoreCursor();

  void reset(boolean clearScrollBackBuffer);

  void characterAttributes(TextStyle textStyle);

  void setScrollingRegion(int top, int bottom);

  void scrollUp(int count);

  void scrollDown(int count);

  void resetScrollRegions();

  void cursorHorizontalAbsolute(int x);

  void linePositionAbsolute(int y);

  void cursorPosition(int x, int y);

  void cursorUp(int countY);

  void cursorDown(int dY);

  void cursorForward(int dX);

  void cursorBackward(int dX);

  void cursorShape(@NotNull CursorShape shape);

  void eraseInLine(int arg);

  void deleteCharacters(int count);

  int getTerminalWidth();

  int getTerminalHeight();

  @NotNull TermSize getSize();

  void eraseInDisplay(int arg);

  void setModeEnabled(TerminalMode mode, boolean enabled);

  void disconnected();

  int getCursorX();

  int getCursorY();

  @NotNull CellPosition getCursorPosition();

  void singleShiftSelect(int num);

  void setWindowTitle(@NotNull String name);

  void saveWindowTitleOnStack();

  void restoreWindowTitleFromStack();

  void clearScreen();

  void setCursorVisible(boolean visible);

  void useAlternateBuffer(boolean enabled);

  byte[] getCodeForKey(int key, int modifiers);

  void setApplicationArrowKeys(boolean enabled);

  void setApplicationKeypad(boolean enabled);

  void setAutoNewLine(boolean enabled);

  StyleState getStyleState();

  void insertLines(int count);

  void deleteLines(int count);

  void eraseCharacters(int count);

  void insertBlankCharacters(int count);

  void clearTabStopAtCursor();

  void clearAllTabStops();

  void setTabStopAtCursor();

  void writeUnwrappedString(String string);

  void setTerminalOutput(@Nullable TerminalOutputStream terminalOutput);

  void setMouseMode(@NotNull MouseMode mode);

  void setMouseFormat(MouseFormat mouseFormat);

  void setAltSendsEscape(boolean enabled);

  void deviceStatusReport(String str);

  void deviceAttributes(byte[] response);

  void setLinkUriStarted(@NotNull String uri);

  void setLinkUriFinished();

  void setBracketedPasteMode(boolean enabled);

  @Nullable Color getWindowForeground();

  @Nullable Color getWindowBackground();

  default void addApplicationTitleListener(@NotNull TerminalApplicationTitleListener listener) {}

  default void removeApplicationTitleListener(@NotNull TerminalApplicationTitleListener listener) {}

  default void addResizeListener(@NotNull TerminalResizeListener listener) {}

  default void removeResizeListener(@NotNull TerminalResizeListener listener) {}

  default void addCustomCommandListener(@NotNull TerminalCustomCommandListener listener) {}

  default void removeCustomCommandListener(@NotNull TerminalCustomCommandListener listener) {}

  default void processCustomCommand(@NotNull List<String> args) {}
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/Terminal.kt
package com.jediterm.terminal

import com.jediterm.core.Color
import com.jediterm.core.util.CellPosition
import com.jediterm.core.util.TermSize
import com.jediterm.terminal.emulator.mouse.MouseFormat
import com.jediterm.terminal.emulator.mouse.MouseMode
import com.jediterm.terminal.model.StyleState
import com.jediterm.terminal.model.TerminalApplicationTitleListener
import com.jediterm.terminal.model.TerminalResizeListener
import java.io.UnsupportedEncodingException

/**
 * Executes terminal commands interpreted by [com.jediterm.terminal.emulator.Emulator], receives text
 *
 * @author traff
 */
interface Terminal {
    fun resize(newTermSize: TermSize, origin: RequestOrigin)

    fun beep()

    fun backspace()

    fun horizontalTab()

    fun carriageReturn()

    fun newLine()

    fun mapCharsetToGL(num: Int)

    fun mapCharsetToGR(num: Int)

    fun designateCharacterSet(tableNumber: Int, ch: Char)

    fun setAnsiConformanceLevel(level: Int)

    @Throws(UnsupportedEncodingException::class)
    fun writeDoubleByte(bytes: CharArray?)

    fun writeCharacters(string: String?)

    fun distanceToLineEnd(): Int

    fun reverseIndex()

    fun index()

    fun nextLine()

    fun fillScreen(c: Char)

    fun saveCursor()

    fun restoreCursor()

    fun reset(clearScrollBackBuffer: Boolean)

    fun characterAttributes(textStyle: TextStyle?)

    fun setScrollingRegion(top: Int, bottom: Int)

    fun scrollUp(count: Int)

    fun scrollDown(count: Int)

    fun resetScrollRegions()

    fun cursorHorizontalAbsolute(x: Int)

    fun linePositionAbsolute(y: Int)

    fun cursorPosition(x: Int, y: Int)

    fun cursorUp(countY: Int)

    fun cursorDown(dY: Int)

    fun cursorForward(dX: Int)

    fun cursorBackward(dX: Int)

    fun cursorShape(shape: CursorShape)

    fun eraseInLine(arg: Int)

    fun deleteCharacters(count: Int)

    fun ambiguousCharsAreDoubleWidth(): Boolean

    val terminalWidth: Int

    val terminalHeight: Int

    val size: TermSize

    fun eraseInDisplay(arg: Int)

    fun setModeEnabled(mode: TerminalMode?, enabled: Boolean)

    fun disconnected()

    val cursorX: Int

    val cursorY: Int

    val cursorPosition: CellPosition

    fun singleShiftSelect(num: Int)

    fun setWindowTitle(name: String)

    fun saveWindowTitleOnStack()

    fun restoreWindowTitleFromStack()

    fun clearScreen()

    fun setCursorVisible(visible: Boolean)

    fun useAlternateBuffer(enabled: Boolean)

    fun getCodeForKey(key: Int, modifiers: Int): ByteArray?

    fun setApplicationArrowKeys(enabled: Boolean)

    fun setApplicationKeypad(enabled: Boolean)

    fun setAutoNewLine(enabled: Boolean)

    val styleState: StyleState?

    fun insertLines(count: Int)

    fun deleteLines(count: Int)

    fun eraseCharacters(count: Int)

    fun insertBlankCharacters(count: Int)

    fun clearTabStopAtCursor()

    fun clearAllTabStops()

    fun setTabStopAtCursor()

    fun writeUnwrappedString(string: String?)

    fun setTerminalOutput(terminalOutput: TerminalOutputStream?)

    fun setMouseMode(mode: MouseMode)

    fun setMouseFormat(mouseFormat: MouseFormat?)

    fun setAltSendsEscape(enabled: Boolean)

    fun deviceStatusReport(str: String?)

    fun deviceAttributes(response: ByteArray?)

    fun setLinkUriStarted(uri: String)

    fun setLinkUriFinished()

    fun setBracketedPasteMode(enabled: Boolean)

    val windowForeground: Color?

    val windowBackground: Color?

    var cursorColor: Color?

    fun addApplicationTitleListener(listener: TerminalApplicationTitleListener) {}

    fun removeApplicationTitleListener(listener: TerminalApplicationTitleListener) {}

    fun addResizeListener(listener: TerminalResizeListener) {}

    fun removeResizeListener(listener: TerminalResizeListener) {}

    fun addCustomCommandListener(listener: TerminalCustomCommandListener) {}

    fun removeCustomCommandListener(listener: TerminalCustomCommandListener) {}

    fun processCustomCommand(args: MutableList<String?>) {}
}
